jfr-connection/src/main/java/io/opentelemetry/contrib/jfr/connection/FlightRecorderDiagnosticCommandConnection.java

```diff
@@ -0,0 +1,243 @@
+import javax.management.MBeanServerConnection;
+import javax.management.MalformedObjectNameException;
+import javax.management.ObjectInstance;
+import javax.management.ObjectName;
+import javax.management.ReflectionException;
+
+/**
+ * This class is the implementation detail of the {@link FlightRecorderConnection} interface which
+ * uses a connection to a {@code com.sun.management DiagnosticCommand} MBean of a JVM. Because it is
+ * implementation detail, it is not part of the public API of this library.
+ *
+ * @see FlightRecorderConnection#diagnosticCommandConnection(MBeanServerConnection)
+ */
+/* package scope */ final class FlightRecorderDiagnosticCommandConnection
+    implements FlightRecorderConnection {
+  private static final String DIAGNOSTIC_COMMAND_OBJECT_NAME =
+      "com.sun.management:type=DiagnosticCommand";
+  private static final String JFR_START_REGEX = "Started recording (.+?)\\. .*";
+  private static final Pattern JFR_START_PATTERN = Pattern.compile(JFR_START_REGEX, Pattern.DOTALL);
+
+  // All JFR commands take String[] parameters
+  private static final String[] signature = new String[] {"[Ljava.lang.String;"};
+
+  /**
+   * Create a connection to the {@code DiagnosticCommand} MBean.
+   *
+   * @param mBeanServerConnection The {@code MBeanServerConnection} to the JVM.
+   * @return A {@code FlightRecorderConnection}.
+   * @throws IOException A communication problem occurred when talking to the MBean server.
+   * @throws JfrConnectionException Indicates an issue with the DiagnosticCommand MBean in the JVM.
+   * @throws NullPointerException The {@code mBeanServerConnection} parameter is {@code null}.
+   * @see FlightRecorderConnection#diagnosticCommandConnection(MBeanServerConnection)
+   */
+  /* package scope */ static FlightRecorderConnection connect(
+      MBeanServerConnection mBeanServerConnection) throws IOException, JfrConnectionException {
+    Objects.requireNonNull(mBeanServerConnection);
+    try {
+      ObjectInstance objectInstance =
+          mBeanServerConnection.getObjectInstance(new ObjectName(DIAGNOSTIC_COMMAND_OBJECT_NAME));
+      ObjectName objectName = objectInstance.getObjectName();
+      assertCommercialFeaturesUnlocked(mBeanServerConnection, objectName);
+      return new FlightRecorderDiagnosticCommandConnection(
+          mBeanServerConnection, objectInstance.getObjectName());
+    } catch (MalformedObjectNameException e) {
+      // Not expected to happen. This exception comes from the ObjectName constructor. If
+      // DIAGNOSTIC_COMMAND_OBJECT_NAME is malformed, then this is an internal bug.
+      throw new JfrConnectionException(DIAGNOSTIC_COMMAND_OBJECT_NAME, e);
+    } catch (InstanceNotFoundException e) {
+      throw JfrConnectionException.canonicalJfrConnectionException(
+          FlightRecorderDiagnosticCommandConnection.class, "connect", e);
+    }
+  }
+
+  /* package scope for testing */ FlightRecorderDiagnosticCommandConnection(
```

Comment:
should that comment go up one line?

Reply:
Yeah, I guess it could. I'll catch this If there are other edits to make.